File: backend/app/api/projects.py
import structlog
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_session
from app.models.project import Project
from app.schemas.project import ProjectCreate, ProjectResponse, ProjectUpdate

logger = structlog.get_logger()
router = APIRouter(prefix="/projects", tags=["projects"])


@router.post("", response_model=ProjectResponse, status_code=status.HTTP_201_CREATED)
async def create_project(
    payload: ProjectCreate,
    session: AsyncSession = Depends(get_session),
) -> Project:
    project = Project(
        name=payload.name,
        duration_hours=payload.duration_hours,
        team_data={
            "team_size": payload.team_size,
            "experience_level": payload.experience_level,
            "skills": payload.skills,
        },
        challenge_data={
            "challenge_statements": payload.challenge_statements,
            "evaluation_criteria": payload.evaluation_criteria,
            "notes": payload.notes,
        },
        resource_data={
            "tracks": payload.tracks,
            "datasets": payload.datasets,
            "apis": payload.apis,
            "documentation_links": payload.documentation_links,
        },
    )
    session.add(project)
    await session.flush()
    await session.refresh(project)
    logger.info("project_created", project_id=project.id, name=project.name)
    return project


@router.get("", response_model=list[ProjectResponse])
async def list_projects(
    session: AsyncSession = Depends(get_session),
) -> list[Project]:
    result = await session.execute(
        select(Project).order_by(Project.created_at.desc())
    )
    return list(result.scalars().all())


@router.get("/{project_id}", response_model=ProjectResponse)
async def get_project(
    project_id: str,
    session: AsyncSession = Depends(get_session),
) -> Project:
    result = await session.execute(
        select(Project).where(Project.id == project_id)
    )
    project = result.scalar_one_or_none()
    if not project:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Project {project_id} not found",
        )
    return project


@router.patch("/{project_id}", response_model=ProjectResponse)
async def update_project(
    project_id: str,
    payload: ProjectUpdate,
    session: AsyncSession = Depends(get_session),
) -> Project:
    result = await session.execute(
        select(Project).where(Project.id == project_id)
    )
    project = result.scalar_one_or_none()
    if not project:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Project {project_id} not found",
        )

    update_data = payload.model_dump(exclude_unset=True)
    for key, value in update_data.items():
        setattr(project, key, value)

    await session.flush()
    await session.refresh(project)
    logger.info("project_updated", project_id=project_id)
    return project


@router.delete("/{project_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_project(
    project_id: str,
    session: AsyncSession = Depends(get_session),
) -> None:
    result = await session.execute(
        select(Project).where(Project.id == project_id)
    )
    project = result.scalar_one_or_none()
    if not project:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Project {project_id} not found",
        )
    await session.delete(project)
    logger.info("project_deleted", project_id=project_id)

File: frontend/src/services/projects.ts
import { api } from "@/lib/api";
import type { HackathonProject, WorkflowProgress } from "@/types";

export interface CreateProjectPayload {
  name: string;
  challenge_statements: string[];
  duration_hours: number;
  team_size: number;
  experience_level: string;
  skills: string[];
  tracks: string[];
  datasets: string[];
  apis: string[];
  documentation_links: string[];
  evaluation_criteria: string[];
  notes?: string;
}

export const projectsService = {
  list: () => api.get<HackathonProject[]>("/projects"),

  get: (id: string) => api.get<HackathonProject>(`/projects/${id}`),

  create: (payload: CreateProjectPayload) =>
    api.post<HackathonProject>("/projects", payload),

  delete: (id: string) => api.delete(`/projects/${id}`),

  startWorkflow: (id: string) =>
    api.post<{ status: string; project_id: string }>(`/workflows/${id}/start`, {}),

  getProgress: (id: string) =>
    api.get<WorkflowProgress>(`/workflows/${id}/progress`),

  runAgent: (projectId: string, agentName: string) =>
    api.post<{ status: string; agent: string }>(
      `/workflows/${projectId}/run-agent/${agentName}`,
      {},
    ),
};
